types/src/on_chain_config.rs

```diff
@@ -7,9 +7,76 @@ use crate::{
     language_storage::{StructTag, TypeTag},
     transaction::SCRIPT_HASH_LENGTH,
 };
+use anyhow::{format_err, Result};
 use libra_crypto::HashValue;
 use move_core_types::identifier::Identifier;
-use serde::{Deserialize, Serialize};
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+use std::collections::HashMap;
+
+/// To register an on-chain config in Rust:
+/// 1. Implement the `OnChainConfig` trait for the Rust representation of the config
+/// 2. Add the config's `ConfigID` to `ON_CHAIN_CONFIG_REGISTRY`
+
+#[derive(Copy, Clone, Eq, Hash, PartialEq)]
+pub struct ConfigID(&'static str);
+
+impl ConfigID {
+    pub fn access_path(self) -> AccessPath {
+        access_path_for_config(Identifier::new(self.0).expect("failed to get Identifier"))
+    }
+}
+
+pub const ON_CHAIN_CONFIG_REGISTRY: &[ConfigID] = &[VMPublishingOption::CONFIG_ID];
+
+#[derive(Clone)]
+pub struct OnChainConfigPayload {
+    configs: HashMap<ConfigID, Option<Vec<u8>>>,
+}
+
+impl OnChainConfigPayload {
+    pub fn new(configs: HashMap<ConfigID, Option<Vec<u8>>>) -> Self {
+        Self { configs }
+    }
+
+    pub fn get<T: OnChainConfig + DeserializeOwned>(&self) -> Result<T> {
+        if let Some(bytes) = self
+            .configs
+            .get(&T::CONFIG_ID)
+            .ok_or_else(|| format_err!("[on-chain-cfg] config not in payload"))?
+        {
+            T::deserialize_into_config(bytes)
+        } else {
+            Err(format_err!("[on-chain-cfg] missing byte array in payload, potentially caused be failed storage read"))
+        }
+    }
+}
+
+/// Trait to be implemented by a storage type from which to read on-chain configs
+pub trait ConfigStorage {
+    fn fetch_config(&self, access_path: AccessPath) -> Option<Vec<u8>>;
```

Comment:
I think eventually we would implement this trait for `LibraDBTrait` too once we replace storage read client with Libra DB and add a sync fetch-config function in `OnChainConfig` trait, so long-term it may be worth keeping this generic?